Fix issues with missing persistence for trade state

chimp1984 opened the pull request:
There are some bug reports which suggest that the persistence at shutdown is not working reliable (hard kill, crash,...).
We reduce the persistence interval radically to ensure important data is written immediately (200 ms delay).
Further the `requestPersistence` has been missing in most trade tasks. I guess that happened when merging the large refactoring projects (trade protocol, persistence). If shutdown routine gets executed it did not had any effect as we wrote then the trade state, but if that was missing we might miss important trade state/data changes.
 
Fixes https://github.com/bisq-network/bisq/issues/4806
Fixes https://github.com/bisq-network/bisq/issues/4762

chimp1984 committed:
Reduce interval for persistence

Seems the persistence at shutdown is too unsafe and we got bug reports where data was missing.
https://github.com/bisq-network/bisq/issues/4806

Use millisec instead of sec for delay
Rename delayInSec to delay

ghost commented:
This would not resolve #4806 since `requestPersistence` for PendingTrades is only called twice: when the trade is new, and when the user closes the trade.  There was > 12 hours in between where the trade progressed through all its states with no call to `requestPersistence`.  

**[EDIT]**  Ok, I see you're trying to persist in SIGTERM.  Maybe it would be ok.  If the delay is only 200ms maybe that is enough time before the process is shut down.  It wouldn't help in the case of power cut though.

chimp1984 commented:
Oh, yes you are right. I though at trade task complete we call it but I fear that got lost in some merge. It was a bit complicate with several larger PRs (refactoring of trade protocol + refactoring of persistence).
I will check how to fix it best...

chimp1984 committed:
Add requestPersistence calls

We relied on the shutdwon routine to be called reliably but it seems that is not the case as some bug reports show.
So we call requestPersistence at every write access of the trade object

chimp1984 commented:
I added the `requestPersistence` at each data change in trade. Hope I did not miss any. We could add some redundancy if we call it at the `complete()` method of each tradeTask but not sure if thats needed.

chimp1984 commented:
Failing test is resolved by @stejbac (https://github.com/bisq-network/bisq/pull/4789#issuecomment-729586729). Will merge in once hie PR is merged.

chimp1984 committed:
Refactoring: Extract requestPersistence method

ghost commented:
ACK, tested, fixes #4806.

chimp1984 committed:
Add more requestPersistence for data changes in ProcessModel and TradingPeer

chimp1984 committed:
Add more requestPersistence calls

chimp1984 commented:
I double checked all trade (and sub models) related fields as well as all trade protocol tasks and protocols. Hope I did not miss a call but as there is quite a lot of redundancy I think its safe now.

sqrrm commented:
I feel it might be cleaner to do the persist in the `complete()` method at the end of each trade task. It feels like that's what's being done with this PR anyway, but separately in each task with the chance of missing it somewhere.

Apart from that I think this looks ok and could be merged.

chimp1984 committed:
Remove requestPersistence in constructor as TradeManager is not set at that moment.

chimp1984 commented on core/src/main/java/bisq/core/trade/protocol/BuyerAsTakerProtocol.java:
```diff
@@ -65,6 +65,7 @@ public BuyerAsTakerProtocol(BuyerAsTakerTrade trade) {
 
         Offer offer = checkNotNull(trade.getOffer());
         processModel.getTradingPeer().setPubKeyRing(offer.getPubKeyRing());
+        processModel.getTradeManager().requestPersistence();
```
Ah in the constructor it must not be called as trademanager is not set yet. I will remove that.

chimp1984 commented on core/src/main/java/bisq/core/trade/protocol/SellerAsTakerProtocol.java:
```diff
@@ -59,6 +59,7 @@ public SellerAsTakerProtocol(SellerAsTakerTrade trade) {
         super(trade);
         Offer offer = checkNotNull(trade.getOffer());
         processModel.getTradingPeer().setPubKeyRing(offer.getPubKeyRing());
+        processModel.getTradeManager().requestPersistence();
```
Ah in the constructor it must not be called as trademanager is not set yet. I will remove that.

chimp1984 commented:
> I feel it might be cleaner to do the persist in the `complete()` method at the end of each trade task. It feels like that's what's being done with this PR anyway, but separately in each task with the chance of missing it somewhere.
> 
> Apart from that I think this looks ok and could be merged.

I was wondering myself if we should add at the tradeTask complete handler a perist call. It would add extra redundancy in case I missed one as well in case we add some tasks or have some changes and the call is forgotten. 
I have no strong opinion for or against it. Signal with up or downvote if you want to have it in.

chimp1984 committed:
Add null checks for tradeManager

We get called some setter methods from protobuf methods before tradeManager is set.

chimp1984 committed:
Handle potential deposit confirmed state issues

The deposit confirmed state is set after we applied the mailbox messages,
which led to a task failure due wrong phase and the message was not applied.
Further it can be that the wallet is still syncing and the deposit
confirmed state is set in any time in the future.

To fix the first problem we add a bit of delay so that the trade has
been updated when we apply the mailbox messages. A better fix would be to change
the order of the methods but that is a bit tricky to get right and I dont want to
risk that for that release.

The second problem would require a large change to trigger the mailbox
processing based on wallet state. We prefer to be more tolerant with
the expected phase instead so allow the mailbox message to be processed
also in the DEPOSIT_PUBLISHED state.
This has no risks as the payout tx would be invalid anyway if the
buyer has cheated and sent the msg in not confirmed deposit tx state (only
possible with code manipulation).

A better fix would to add a listener for the wallet and process
the mailbox msg once wallet is ready and trade state set, but I
leave that for another PR.

chimp1984 committed:
Add requestPersistence call at TradeTask.complete call.

This is not really needed as we call it at each state change of the
trade but gives more redundancy in case we missed one or once
changes are applied and a dev forgets to call it.

Multiple repeated calls do have close to zero costs.

chimp1984 committed:
Move delay for applying mailbox messages inside onInitialized

We need to set addDecryptedDirectMessageListener without
delay as otherwise we could miss direct messages (detected
with localhost testing, with tor its likely slower and
would not have been triggered).

sqrrm commented:
With the persistence call in complete() it's not necessary to persist in each trade task.

chimp1984 commented:
> With the persistence call in complete() it's not necessary to persist in each trade task.

A reason why I did not do want to rely only on complete is as there are async tasks and i want to be sure that if the task never completes (shutdown) that the intermediate state change is persisted. But also not happy that its that verbose now ;-(. But its more similar as it was before, just that some calls have been done inside the trade setter methods.

sqrrm commented:
Agreed that some persist calls should still be done in the tasks, but all of those done synchronously just before complete() are no longer needed.
